### PR TITLE
Reduce allocations in CookieContainer.GetCookieHeader

### PR DESCRIPTION
**Pull request by stephentoub**
HttpClient defaults to UseCookies=true, which means by default every request ends up asking the CookieContainer to get the cookie header that should be used. Today, even if there aren't any cookies for the given Uri, this ends up doing a lot of allocation.  And if there are cookies for the Uri, the serialization of each cookie results in a handful of strings, char[]s, and string[] allocations. This commit tweaks the implementation to avoid a bunch of these, including taking advantage of StringBuilderCache. (There are still some more that could be cleaned up with further work, but it becomes much more invasive.)

For this sample program:
```C#
CookieContainer cc = new CookieContainer();

Uri exists = new Uri("http://example.com/exists");
Uri notExists = new Uri("http://example.com/notExists");

cc.Add(exists, new Cookie("test1", "value1"));
cc.Add(exists, new Cookie("test2", "value2"));

for (int i = 0; i < 1000; i++)
{
    cc.GetCookieHeader(exists);
}
```
before these changes, I see the following allocations:
image
and after thes changes:
image

If I then change the GetCookieHeader line to be:
```C#
cc.GetCookieHeader(notExists);
```
before these changes, I see the following allocations:
image
and after these changes:
image

cc: @davidsh, @cipop, @ericeil

### CODE REVIEW

File: src/System.Net.Primitives/src/System/Net/Cookie.cs [hunk 1]
```diff
@@ -2,10 +2,11 @@
 // The .NET Foundation licenses this file to you under the MIT license.
```
Comment (davidsh) [inline src/System.Net.Primitives/src/System/Net/Cookie.cs]: You should consider making equivalent changes to the copy of Cookie.cs here:
https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/netcore50/System/Net/cookie.cs

Long term....there won't be a copy of the file.  But due to the way that netcore50 needs cookie functionality, it needed to duplicate the file for now....could potentially be refactored to common code, etc.

**Comment by stephentoub** [comment]
@dotnet-bot test this please (Jenkins was down)

**Comment by stephentoub** [comment]
Test Innerloop Ubuntu Debug Build and Test please (timeout during package restore)

**Comment by stephentoub** [comment]
Thanks, @davidsh.  I ported the Cookie changes over to cookie.cs.  I decided not to use StringBuilderCache (yet), since it's not currently used in the Http library, but that's something we likely want to do for other reasons soon.

**Comment by davidsh** [comment]
LGTM

**Comment by stephentoub** [comment]
Thanks, @davidsh.  Do you want to run any .NET Native tests before this is merged?

**Comment by davidsh** [comment]
I'm ok w/ merging now.

**Pull request merged**